components/common/src/liveliness_checker.rs

```diff
@@ -0,0 +1,467 @@
+    static ref THREAD_STATUSES: Mutex<ThreadStatusMap> = Default::default();
+}
+
+habitat_core::env_config_duration!(ThreadAliveThreshold,
+                                       HAB_THREAD_ALIVE_THRESHOLD_SECS => from_secs,
+                                       Duration::from_secs(5 * 60));
+habitat_core::env_config_duration!(ThreadAliveCheckDelay,
+                                       HAB_THREAD_ALIVE_CHECK_DELAY_SECS => from_secs,
+                                       Duration::from_secs(60));
+habitat_core::env_config_duration!(ThreadDeadIgnoreDelay,
+                                       HAB_THREAD_DEAD_IGNORE_DELAY_SECS => from_secs,
+                                       Duration::from_secs(60 * 60 * 24 * 7)); // 1 week
+
+/// Call periodically from a thread which has a work loop to indicate that the thread is
+/// still alive and processing its loop. If this function is not called for more than
+/// `ThreadAliveThreshold`, it will be error logged as a likely deadlock.
+///
+/// If the thread exits the loop where mark_thread_alive will be called, it must be unregistered
+/// from the liveliness checker or else be subject to false positives. In general, the idiom is
+/// ```
+/// use habitat_common::liveliness_checker;
+/// # let error = false;
+/// let _: liveliness_checker::ThreadUnregistered<_, _> = loop {
+///     let checked_thread = liveliness_checker::mark_thread_alive();
+///     if error {
+///         break checked_thread.unregister(Err("some description"));
+///     } else {
+///         break checked_thread.unregister(Ok(()));
+///     }
+/// };
+/// ```
+pub fn mark_thread_alive() -> CheckedThread {
+    mark_thread_alive_impl(&mut THREAD_STATUSES.lock().expect("THREAD_STATUSES poisoned"))
+}
+
+fn mark_thread_alive_impl(statuses: &mut ThreadStatusMap) -> CheckedThread {
+    let thread = thread::current();
+    statuses.insert(thread.id(),
+                    (thread.name().map(str::to_string),
+                     Status::Alive { last_heartbeat: Instant::now(), }));
+    CheckedThread(std::ptr::null())
+}
+
+/// A type to enforce that all code paths to exit a thread call unregister_thread. Since this
+/// type can't be instantiated outside this module, the only way to have consistent return
+/// types is to call unregister_thread on all returning paths. This can't prevent the error
+/// that *no* code paths call unregister_thread, but it should help avoid the situation of a
+/// particular one being overlooked.
+///
+/// Additionally, since code paths which the threads pass through while unregistering may wish
+/// to communicate a Result back to the caller, this type wraps an arbitrary `Result` which can
+/// be then recovered via `into_result`, or if an arbitrary value is desired, via `into_ok`.
+#[must_use]
+pub struct ThreadUnregistered<T = (), E = std::convert::Infallible>(Result<T, E>);
+
+impl<T, E> ThreadUnregistered<T, E> {
+    pub fn into_result(self) -> Result<T, E> { self.0 }
```

Comment:
In case you're wondering why we can't use `Into`/`From` traits: https://users.rust-lang.org/t/generic-conversion-from-newtypes/16247